### PR TITLE
Startup test

### PR DESCRIPTION
**Pull request by IdoZilberberg**

### CODE REVIEW

File: deploy/bootstrap/docker-compose.yml [hunk 1]
```diff
@@ -45,7 +45,8 @@ services:
 
   public-api:
     image: ${DOCKER_IMAGE}:${DOCKER_TAG}
-    working_dir: /opt/orbs/projects/services/public-api-service-typescript/
+    # working_dir: /opt/orbs/projects/services/public-api-service-typescript/
```
Comment (netoneko) [inline deploy/bootstrap/docker-compose.yml]: It won't work. Please revert this change.

File: Dockerfile.server [hunk 2]
```diff
@@ -3,3 +3,8 @@ FROM orbs:base-server
 ADD . /opt/orbs
 
 RUN ./build-server.sh && yarn cache clean
+
+HEALTHCHECK --interval=10s --timeout=10s --retries=12 \
+    CMD curl -q http://localhost:8081/admin/startupCheck
+
+#CMD ./entrypoint.sh
```
Comment (netoneko) [inline Dockerfile.server]: There is no `entrypoint.sh` and this piece is commented out.

File: idea-codestyle.xml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+<code_scheme name="Default" version="173">
```
Comment (netoneko) [inline idea-codestyle.xml]: Why do we need this file?

File: projects/libs/core-library-typescript/package.json [hunk 4]
```diff
@@ -72,6 +75,7 @@
     "crc": "^3.5.0",
     "ganache-core": "^2.0.2",
     "mocha": "^5.0.0",
+    "request-promise": "^4.2.2",
```
Comment (netoneko) [inline projects/libs/core-library-typescript/package.json]: Where do we use it?

File: projects/libs/core-library-typescript/src/block-storage/block-storage.ts [hunk 5]
```diff
@@ -131,4 +136,14 @@ export class BlockStorage {
...
     await this.db.put<string>(block.header.height.toString(), JSON.stringify(block));
   }
+
+  public async startupCheck(): Promise<StartupStatus> {
+
+    if (!this.transactionPool) {
+      return <StartupStatus>{ name: this.SERVICE_NAME, status: STARTUP_STATUS.FAIL };
+    }
+
+    const lastBlock = await this.getLastBlock();
```
Comment (netoneko) [inline projects/libs/core-library-typescript/src/block-storage/block-storage.ts]: Needs to be in try/catch because if `getLastBlock` fails we want to return an error.

File: projects/libs/core-library-typescript/src/block-storage/block-storage.ts [hunk 6]
```diff
@@ -131,4 +136,14 @@ export class BlockStorage {
...
+    return <StartupStatus>{ name: this.SERVICE_NAME, status: lastBlock ? STARTUP_STATUS.OK : STARTUP_STATUS.FAIL };
```
Comment (netoneko) [inline projects/libs/core-library-typescript/src/block-storage/block-storage.ts]: Unnecessary cast.

File: projects/libs/core-library-typescript/src/common-library/grpc-server.ts [hunk 7]
```diff
@@ -51,19 +69,54 @@ export class GRPCServerBuilder {
...
+
+    if (!this.managementPort) {
+      return Promise.reject("No management port defined, you must call withManagementPort() before starting");
+    }
+
+    if (!this.mali) {
+      return Promise.reject("Mali was not set up correctly. did you forget to call withService()?");
+    }
+
+    // TODO extract to method
```
Comment (netoneko) [inline projects/libs/core-library-typescript/src/common-library/grpc-server.ts]: Extract it outside, this code should be in a separate file.

File: projects/libs/core-library-typescript/src/common-library/grpc-server.ts [hunk 8]
```diff
@@ -51,19 +69,54 @@ export class GRPCServerBuilder {
...
+
+      });
   }
 
   stop(): Promise<any> {
     const all = Promise.all(this.services.map(s => s.stop()));
-
-    this.mali.close(this.endpoint);
-
+    this.mali && this.mali.close(this.endpoint);
```
Comment (netoneko) [inline projects/libs/core-library-typescript/src/common-library/grpc-server.ts]: Please use if statement here.

File: projects/libs/core-library-typescript/src/common-library/startup-check-runner.ts [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    const startupCheckPromises = this.startupCheckers.map((s: StartupCheck) => s.startupCheck());
+    return Promise.all(startupCheckPromises)
+      .then(this.mergeStartupStatuses)
+      .catch(err => {
+        logger.error(err);
+        return <StartupStatus>{ status: STARTUP_STATUS.FAIL, message: err.message };
+      });
+  }
+
+  private mergeStartupStatuses = (startupStatuses: StartupStatus[]): StartupStatus => {
```
Comment (netoneko) [inline projects/libs/core-library-typescript/src/common-library/startup-check-runner.ts]: Extract as a method.

File: projects/libs/core-library-typescript/src/common-library/startup-status.ts [hunk 10]
```diff
@@ -0,0 +1,14 @@
+export interface StartupStatus {
+  name: string;
+  status: STARTUP_STATUS;
+  childStartupStatuses?: StartupStatus[];
+  message?: string;
+}
+
+// export interface CompositeStartupStatus {
```
Comment (netoneko) [inline projects/libs/core-library-typescript/src/common-library/startup-status.ts]: Please remove this.

File: projects/libs/core-library-typescript/src/state-storage/state-storage.ts [hunk 11]
```diff
@@ -6,8 +6,11 @@ import { types } from "../common-library/types";
 import { InMemoryKVStore } from "./kvstore";
 
 import { delay } from "bluebird";
+import { STARTUP_STATUS, StartupStatus } from "../common-library/startup-status";
+import { StartupCheck } from "../common-library/startup-check";
 
-export class StateStorage {
+export class StateStorage implements StartupCheck {
+  private SERVICE_NAME = "state";
```
Comment (netoneko) [inline projects/libs/core-library-typescript/src/state-storage/state-storage.ts]: `state-storage`

File: projects/libs/core-library-typescript/src/state-storage/state-storage.ts [hunk 12]
```diff
@@ -94,12 +97,30 @@ export class StateStorage {
...
     } else {
-      throw new Error(`Unexpected block Height: ${block.header.height}. Expected ${ this.lastBlockHeight + 1 }, Out of sync?`);
+      throw new Error(`Unexpected block Height: ${block.header.height}. Expected ${this.lastBlockHeight + 1}, Out of sync?`);
     }
   }
+
+  public async startupCheck(): Promise<StartupStatus> {
+
+    if (!this.kvstore) {
+      return <StartupStatus>{ name: this.SERVICE_NAME, status: STARTUP_STATUS.FAIL, message: "Missing kvstore" };
```
Comment (netoneko) [inline projects/libs/core-library-typescript/src/state-storage/state-storage.ts]: Unnecessary casts.

File: projects/libs/core-library-typescript/src/state-storage/state-storage.ts [hunk 13]
```diff
@@ -94,12 +97,30 @@ export class StateStorage {
...
+    }
+
+    if (!this.pollIntervalMs) {
```
Comment (netoneko) [inline projects/libs/core-library-typescript/src/state-storage/state-storage.ts]: No need to check this. It's already initialized in the constructor.

File: projects/libs/core-library-typescript/src/state-storage/state-storage.ts [hunk 14]
```diff
@@ -94,12 +97,30 @@ export class StateStorage {
...
+      return <StartupStatus>{ name: this.SERVICE_NAME, status: STARTUP_STATUS.FAIL, message: "Missing pollIntervalMs" };
+    }
+
+    if (!this.blockStorage) {
+      return <StartupStatus>{ name: this.SERVICE_NAME, status: STARTUP_STATUS.FAIL, message: "Missing blockStorage" };
```
Comment (netoneko) [inline projects/libs/core-library-typescript/src/state-storage/state-storage.ts]: "Missing block storage configuration" or "Missing block storage client".

File: projects/libs/core-library-typescript/test/block-storage/block-storage.spec.ts [hunk 15]
```diff
@@ -117,4 +121,33 @@ describe("Block storage", () => {
       await expect(blockStorage.getBlocks(lastBlock.header.height)).to.eventually.be.eql([exampleBlock]);
     });
   });
+
+  describe("run checkStartupStatus", () => {
+    it("checkStartupStatus should succeed", async () => {
+      const startupStatus = await blockStorage.startupCheck();
+      return expect(startupStatus).to.deep.equal(<StartupStatus>{ name: "block", status: STARTUP_STATUS.OK });
```
Comment (netoneko) [inline projects/libs/core-library-typescript/test/block-storage/block-storage.spec.ts]: Please remove these casts from the codebase as a whole.

File: projects/libs/core-library-typescript/test/block-storage/block-storage.spec.ts [hunk 16]
```diff
@@ -117,4 +121,33 @@ describe("Block storage", () => {
...
+    });
+  });
+});
+
+describe("Block storage - Bad setup (no transaction pool)", () => {
+  let blockStorage: BlockStorage;
+  // let lastBlock: types.Block;
```
Comment (netoneko) [inline projects/libs/core-library-typescript/test/block-storage/block-storage.spec.ts]: Please remove all commented-out stuff from this pull request.

File: projects/libs/core-library-typescript/test/block-storage/block-storage.spec.ts [hunk 17]
```diff
@@ -117,4 +121,33 @@ describe("Block storage", () => {
...
+
+  beforeEach(async () => {
+    try {
+      fsExtra.removeSync(LEVELDB_PATH);
+    } catch (e) { }
+    blockStorage = await initBlockStorage(undefined);
```
Comment (netoneko) [inline projects/libs/core-library-typescript/test/block-storage/block-storage.spec.ts]: Why do you init with `undefined`?

File: projects/libs/core-library-typescript/test/common-library/grpc-server.spec.ts [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+const { expect } = chai;
+chai.use(chaiAsPromised);
+
+describe("GRPCServer cannot start if no services or no endpoint are attached to it", () => {
+
+  it("should Promise.reject() if trying to start with no endpoint", async () => {
+    // This endpoint is not used, it's just so GRPCserver will not fail on missing endpoint
+    const builder = new GRPCServerBuilder().withManagementPort(8081);
+    const expectation = expect(builder.start()).to.be.rejected;
+    if (builder) {
```
Comment (netoneko) [inline projects/libs/core-library-typescript/test/common-library/grpc-server.spec.ts]: Please use `let builder` and `afterEach()` combo.

File: projects/libs/core-library-typescript/src/common-library/startup-status.ts [hunk 19]
```diff
@@ -0,0 +1,14 @@
+export interface StartupStatus {
+  name: string;
+  status: STARTUP_STATUS;
+  childStartupStatuses?: StartupStatus[];
```
Comment (netoneko) [inline projects/libs/core-library-typescript/src/common-library/startup-status.ts]: Please replace with `services`.

**Comment by netoneko** [comment]
Please also log the results of startup checks as `logger.info` on service level.

File: Dockerfile.server (hunk 2 above)
Comment (IdoZilberberg) [reply Dockerfile.server]: Removed

File: idea-codestyle.xml (hunk 3 above)
Comment (IdoZilberberg) [reply idea-codestyle.xml]: It was meant for IntelliJ IDEA, but for now we won't be using it. Removed.

File: projects/libs/core-library-typescript/package.json (hunk 4 above)
Comment (IdoZilberberg) [reply projects/libs/core-library-typescript/package.json]: Removed express,request-promise from devDependencies.
rm -rf dist & rm -rf distTree are useful to remove junk on build and they are very fast.

File: projects/libs/core-library-typescript/src/block-storage/block-storage.ts (hunk 5 above)
Comment (IdoZilberberg) [reply projects/libs/core-library-typescript/src/block-storage/block-storage.ts]: Added

File: projects/libs/core-library-typescript/src/block-storage/block-storage.ts (hunk 6 above)
Comment (IdoZilberberg) [reply projects/libs/core-library-typescript/src/block-storage/block-storage.ts]: Removed all similar casts

File: projects/libs/core-library-typescript/src/common-library/grpc-server.ts (hunk 8 above)
Comment (IdoZilberberg) [reply projects/libs/core-library-typescript/src/common-library/grpc-server.ts]: Ok, although this is identical to "if", and is just as readable. 
I guess for V1 we may want to discuss code style more thoroughly. I'm fine with either method.

File: projects/libs/core-library-typescript/src/common-library/startup-check-runner.ts (hunk 9 above)
Comment (IdoZilberberg) [reply projects/libs/core-library-typescript/src/common-library/startup-check-runner.ts]: What is the problem with this impl?

File: projects/libs/core-library-typescript/src/common-library/startup-status.ts (hunk 19 above)
Comment (IdoZilberberg) [reply projects/libs/core-library-typescript/src/common-library/startup-status.ts]: This might not always relate just to services. Had a discussion with Shai about it.

File: projects/libs/core-library-typescript/src/state-storage/state-storage.ts (hunk 11 above)
Comment (IdoZilberberg) [reply projects/libs/core-library-typescript/src/state-storage/state-storage.ts]: Changed "state" to "state-storage" and "block" to "block-storage".

File: projects/libs/core-library-typescript/src/state-storage/state-storage.ts (hunk 12 above)
Comment (IdoZilberberg) [reply projects/libs/core-library-typescript/src/state-storage/state-storage.ts]: Removed.

File: projects/libs/core-library-typescript/src/state-storage/state-storage.ts (hunk 13 above)
Comment (IdoZilberberg) [reply projects/libs/core-library-typescript/src/state-storage/state-storage.ts]: This is a defensive check - what if someone decides to remove this from ctor..

File: projects/libs/core-library-typescript/src/state-storage/state-storage.ts (hunk 14 above)
Comment (IdoZilberberg) [reply projects/libs/core-library-typescript/src/state-storage/state-storage.ts]: Changed to "Missing block storage client"

File: projects/libs/core-library-typescript/test/block-storage/block-storage.spec.ts (hunk 15 above)
Comment (IdoZilberberg) [reply projects/libs/core-library-typescript/test/block-storage/block-storage.spec.ts]: Removed all except where is causes errors such as startup-check-runner.ts

File: projects/libs/core-library-typescript/test/common-library/grpc-server.spec.ts (hunk 18 above)
Comment (IdoZilberberg) [reply projects/libs/core-library-typescript/test/common-library/grpc-server.spec.ts]: Fixed

File: projects/libs/core-library-typescript/test/block-storage/block-storage.spec.ts (hunk 17 above)
Comment (IdoZilberberg) [reply projects/libs/core-library-typescript/test/block-storage/block-storage.spec.ts]: On purpose, to init an invalid value so a test can fail

File: projects/libs/core-library-typescript/src/common-library/startup-check-runner.ts (hunk 9 above)
Comment (netoneko) [reply projects/libs/core-library-typescript/src/common-library/startup-check-runner.ts]: It's a function defined inside a class and not a method, which is very unusual and does not serve any practical purpose.

File: projects/libs/core-library-typescript/src/common-library/startup-status.ts (hunk 19 above)
Comment (netoneko) [reply projects/libs/core-library-typescript/src/common-library/startup-status.ts]: Shai does not define the requirements for this task. His input is valuable but I specifically asked for a certain API which we discussed in the beginning of last week.

Comment (IdoZilberberg) [reply projects/libs/core-library-typescript/src/common-library/startup-status.ts]: You mean replace childStartupStatuses with services?

**Comment by IdoZilberberg** [comment]
To prevent failures of startup check due to unimplemented checks on some services, there is a default startup checker for services which do not explicitly define a startup checker.
The default returns status: "OK" with message "Not implemented".

File: Dockerfile.server (hunk 2 above)
Comment (IdoZilberberg) [reply Dockerfile.server]: Now I feel you mean that you want to have entrypoint.sh, so in that case let's talk tomorrow about what it should contain.

File: projects/libs/core-library-typescript/src/common-library/startup-check-runner.ts [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+    this.startupCheckers = startupCheckers;
+  }
+
+  run(): Promise<StartupStatus> {
+
+    const startupCheckPromises = this.startupCheckers.map((s: StartupCheck) => s.startupCheck());
+    return Promise.all(startupCheckPromises)
+      .then(statuses => {
+        const mergedStartupStatus = this.mergeStartupStatuses(this.name, statuses);
+        logger.info(`StartupCheckRunner.run() result: ${JSON.stringify(mergedStartupStatus)}`);
```
Comment (netoneko) [inline projects/libs/core-library-typescript/src/common-library/startup-check-runner.ts]: `logger.info("Health check", mergedStartupStatus);`

Comment (IdoZilberberg) [reply projects/libs/core-library-typescript/src/common-library/startup-check-runner.ts]: Done.
